In [26]:
import numpy as np
import pandas as pd
import plotly.express as px 

#### 1. Load raw data

In [27]:
df_raw = pd.read_csv('../data/raw/books_1.Best_Books_Ever.csv')

#### 2. Basic Overview

In [28]:
df_raw.head()

,bookId,title,series,author,rating,description,language,isbn,genres,characters,...,firstPublishDate,awards,numRatings,ratingsByStars,likedPercent,setting,coverImg,bbeScore,bbeVotes,price
0,2767052-the-hunger-games,The Hunger Games,The Hunger Games #1,Suzanne Collins,4.33,WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...,English,9780439023481,"['Young Adult', 'Fiction', 'Dystopia', 'Fantas...","['Katniss Everdeen', 'Peeta Mellark', 'Cato (H...",...,NaN,['Locus Award Nominee for Best Young Adult Boo...,6376780,"['3444695', '1921313', '745221', '171994', '93...",96.0,"['District 12, Panem', 'Capitol, Panem', 'Pane...",https://i.gr-assets.com/images/S/compressed.ph...,2993816,30516,5.09
1,2.Harry_Potter_and_the_Order_of_the_Phoenix,Harry Potter and the Order of the Phoenix,Harry Potter #5,"J.K. Rowling, Mary GrandPré (Illustrator)",4.50,There is a door at the end of a silent corrido...,English,9780439358071,"['Fantasy', 'Young Adult', 'Fiction', 'Magic',...","['Sirius Black', 'Draco Malfoy', 'Ron Weasley'...",...,06/21/03,['Bram Stoker Award for Works for Young Reader...,2507623,"['1593642', '637516', '222366', '39573', '14526']",98.0,['Hogwarts School of Witchcraft and Wizardry (...,https://i.gr-assets.com/images/S/compressed.ph...,2632233,26923,7.38
2,2657.To_Kill_a_Mockingbird,To Kill a Mockingbird,To Kill a Mockingbird,Harper Lee,4.28,The unforgettable novel of a childhood in a sl...,English,9999999999999,"['Classics', 'Fiction', 'Historical Fiction', ...","['Scout Finch', 'Atticus Finch', 'Jem Finch', ...",...,07/11/60,"['Pulitzer Prize for Fiction (1961)', 'Audie A...",4501075,"['2363896', '1333153', '573280', '149952', '80...",95.0,"['Maycomb, Alabama (United States)']",https://i.gr-assets.com/images/S/compressed.ph...,2269402,23328,NaN
3,1885.Pride_and_Prejudice,Pride and Prejudice,NaN,"Jane Austen, Anna Quindlen (Introduction)",4.26,Alternate cover edition of ISBN 9780679783268S...,English,9999999999999,"['Classics', 'Fiction', 'Romance', 'Historical...","['Mr. Bennet', 'Mrs. Bennet', 'Jane Bennet', '...",...,01/28/13,[],2998241,"['1617567', '816659', '373311', '113934', '767...",94.0,"['United Kingdom', 'Derbyshire, England (Unite...",https://i.gr-assets.com/images/S/compressed.ph...,1983116,20452,NaN
4,41865.Twilight,Twilight,The Twilight Saga #1,Stephenie Meyer,3.60,About three things I was absolutely positive.\...,English,9780316015844,"['Young Adult', 'Fantasy', 'Romance', 'Vampire...","['Edward Cullen', 'Jacob Black', 'Laurent', 'R...",...,10/05/05,"['Georgia Peach Book Award (2007)', 'Buxtehude...",4964519,"['1751460', '1113682', '1008686', '542017', '5...",78.0,"['Forks, Washington (United States)', 'Phoenix...",https://i.gr-assets.com/images/S/compressed.ph...,1459448,14874,2.1


In [29]:
df_raw.describe()

,rating,numRatings,likedPercent,bbeScore,bbeVotes
count,52478.000000,5.247800e+04,51856.000000,5.247800e+04,52478.000000
mean,4.021878,1.787865e+04,92.231545,1.984023e+03,22.529003
std,0.367146,1.039448e+05,5.990689,3.515314e+04,369.158541
min,0.000000,0.000000e+00,0.000000,0.000000e+00,-4.000000
25%,3.820000,3.410000e+02,90.000000,8.400000e+01,1.000000
50%,4.030000,2.307000e+03,94.000000,9.700000e+01,1.000000
75%,4.230000,9.380500e+03,96.000000,1.870000e+02,2.000000
max,5.000000,7.048471e+06,100.000000,2.993816e+06,30516.000000


In [30]:
df_raw.shape

(52478, 25)

#### 3. Schema Inspection

In [31]:
df_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 52478 entries, 0 to 52477
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   bookId            52478 non-null  str    
 1   title             52478 non-null  str    
 2   series            23470 non-null  str    
 3   author            52478 non-null  str    
 4   rating            52478 non-null  float64
 5   description       51140 non-null  str    
 6   language          48672 non-null  str    
 7   isbn              52478 non-null  str    
 8   genres            52478 non-null  str    
 9   characters        52478 non-null  str    
 10  bookFormat        51005 non-null  str    
 11  edition           4955 non-null   str    
 12  pages             50131 non-null  str    
 13  publisher         48782 non-null  str    
 14  publishDate       51598 non-null  str    
 15  firstPublishDate  31152 non-null  str    
 16  awards            52478 non-null  str    
 17  numR

The dataset contains 52478 rows and 25 columns. 20 columns are categorical and 5 are numeric. Genres, characters, awards, ratingsbystarts and setting are high-cardinality columns. Bookid is the ID column of the dataset. Some columns contain lists but have a Dtype of str.

In [32]:
schema_audit = pd.DataFrame({
    "column": df_raw.columns,
    "dtype": df_raw.dtypes.astype(str).values,
    "non_null_count": df_raw.notna().sum().values,
    "missing_count": df_raw.isna().sum().values,
    "missing_percent": (df_raw.isna().mean() * 100).round(2).values,
    "unique_values": [df_raw[col].nunique(dropna=True) for col in df_raw.columns]
})

schema_audit.sort_values("missing_percent", ascending=False)

,column,dtype,non_null_count,missing_count,missing_percent,unique_values
11,edition,str,4955,47523,90.56,1788
2,series,str,23470,29008,55.28,22802
15,firstPublishDate,str,31152,21326,40.64,8024
24,price,str,38113,14365,27.37,3776
6,language,str,48672,3806,7.25,81
13,publisher,str,48782,3696,7.04,11110
12,pages,str,50131,2347,4.47,1365
10,bookFormat,str,51005,1473,2.81,136
5,description,str,51140,1338,2.55,50888
14,publishDate,str,51598,880,1.68,10635


#### 4. Missingness audit

In [33]:
missing = df_raw.isnull().mean().sort_values(ascending=False).mul(100).round(2)
missing

edition             90.56
series              55.28
firstPublishDate    40.64
price               27.37
language             7.25
publisher            7.04
pages                4.47
bookFormat           2.81
description          2.55
publishDate          1.68
likedPercent         1.19
coverImg             1.15
isbn                 0.00
ratingsByStars       0.00
bbeVotes             0.00
bbeScore             0.00
setting              0.00
awards               0.00
numRatings           0.00
genres               0.00
author               0.00
rating               0.00
title                0.00
characters           0.00
bookId               0.00
dtype: float64

The highest missingness appears in `edition`, `series`, `firstPublishDate`, and `price`.

These missing values should not be handled the same way:
- `edition` is likely optional metadata and may not be useful for core analysis.
- `series` missingness may be meaningful because many books are standalone.
- `firstPublishDate` is important for temporal analysis but has substantial missingness.
- `price` needs further investigation before deciding whether to use it.

#### 5. Duplicate Audit

In [34]:
exact_duplicate_count = df_raw.duplicated().sum()
exact_duplicate_count

np.int64(50)

In [35]:
exact_duplicates = df_raw[df_raw.duplicated(keep=False)].sort_values("bookId")
exact_duplicates.head(10)

,bookId,title,series,author,rating,description,language,isbn,genres,characters,...,firstPublishDate,awards,numRatings,ratingsByStars,likedPercent,setting,coverImg,bbeScore,bbeVotes,price
37460,11737313-three-times-lucky,Three Times Lucky,Mo & Dale Mysteries #1,Sheila Turnage,4.02,"Newbery honor winner, New York Times bestselle...",English,9780803736702,"['Mystery', 'Middle Grade', 'Fiction', 'Realis...",[],...,NaN,"['Newbery Medal Nominee (2013)', ""Pennsylvania...",22422,"['8633', '7941', '4181', '1076', '591']",93.0,['North Carolina (United States)'],https://i.gr-assets.com/images/S/compressed.ph...,87,1,3.06
37406,11737313-three-times-lucky,Three Times Lucky,Mo & Dale Mysteries #1,Sheila Turnage,4.02,"Newbery honor winner, New York Times bestselle...",English,9780803736702,"['Mystery', 'Middle Grade', 'Fiction', 'Realis...",[],...,NaN,"['Newbery Medal Nominee (2013)', ""Pennsylvania...",22422,"['8633', '7941', '4181', '1076', '591']",93.0,['North Carolina (United States)'],https://i.gr-assets.com/images/S/compressed.ph...,87,1,3.06
37442,1258333.Cantarella_Volume_1,"Cantarella, Volume 1",Cantarella #1,You Higuri,3.81,"From birth, Cesare Borgia is surrounded by sha...",English,9780976895701,"['Manga', 'Historical Fiction', 'Fantasy', 'Gr...","['Cesare Borgia', 'Lucrezia Borgia', 'Pope Ale...",...,03/08/01,[],933,"['324', '250', '253', '71', '35']",89.0,['Italy'],https://i.gr-assets.com/images/S/compressed.ph...,87,1,4.67
37388,1258333.Cantarella_Volume_1,"Cantarella, Volume 1",Cantarella #1,You Higuri,3.81,"From birth, Cesare Borgia is surrounded by sha...",English,9780976895701,"['Manga', 'Historical Fiction', 'Fantasy', 'Gr...","['Cesare Borgia', 'Lucrezia Borgia', 'Pope Ale...",...,03/08/01,[],933,"['324', '250', '253', '71', '35']",89.0,['Italy'],https://i.gr-assets.com/images/S/compressed.ph...,87,1,4.67
37428,13571757-1,アオハライド 1,アオハライド [Ao Haru Ride] #1,"Io Sakisaka, 咲坂伊緒",4.33,高1の双葉には中学時代から忘れられない初恋の人、「田中くん」がいる。中学のとき女子に、はぶら...,Japanese,9784088466477,"['Manga', 'Romance', 'Young Adult', 'Shojo', '...",[],...,NaN,[],16361,"['9121', '4447', '2102', '451', '240']",96.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,13.30
37482,13571757-1,アオハライド 1,アオハライド [Ao Haru Ride] #1,"Io Sakisaka, 咲坂伊緒",4.33,高1の双葉には中学時代から忘れられない初恋の人、「田中くん」がいる。中学のとき女子に、はぶら...,Japanese,9784088466477,"['Manga', 'Romance', 'Young Adult', 'Shojo', '...",[],...,NaN,[],16361,"['9121', '4447', '2102', '451', '240']",96.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,13.30
37461,13614864-know-not-why,Know Not Why,Know Not Why #1,Hannah Johnson (Goodreads Author),4.02,Howie gets a job at Artie Kraft's Arts 'N Craf...,English,9781476149035,"['M M Romance', 'LGBT', 'Romance', 'Contempora...","[""Howard 'Howie' Jenkins"", 'Arthur Kraft Jr.',...",...,NaN,[],3411,"['1290', '1231', '632', '189', '69']",92.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,NaN
37407,13614864-know-not-why,Know Not Why,Know Not Why #1,Hannah Johnson (Goodreads Author),4.02,Howie gets a job at Artie Kraft's Arts 'N Craf...,English,9781476149035,"['M M Romance', 'LGBT', 'Romance', 'Contempora...","[""Howard 'Howie' Jenkins"", 'Arthur Kraft Jr.',...",...,NaN,[],3411,"['1290', '1231', '632', '189', '69']",92.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,NaN
37483,1481056.Gifts_and_Ministries_of_the_Holy_Spirit,Gifts and Ministries of the Holy Spirit,NaN,Lester Sumrall,4.44,You Can Have the Power of the Holy Spirit The...,English,9780883686522,"['Christian', 'Theology', 'Christian Non Ficti...",[],...,06/06/77,[],333,"['206', '83', '31', '11', '2']",96.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,5.00
37429,1481056.Gifts_and_Ministries_of_the_Holy_Spirit,Gifts and Ministries of the Holy Spirit,NaN,Lester Sumrall,4.44,You Can Have the Power of the Holy Spirit The...,English,9780883686522,"['Christian', 'Theology', 'Christian Non Ficti...",[],...,06/06/77,[],333,"['206', '83', '31', '11', '2']",96.0,[],https://i.gr-assets.com/images/S/

In [36]:
bookid_duplicate_count = df_raw.duplicated(subset=["bookId"]).sum()
bookid_duplicate_count

np.int64(54)

In [37]:
title_author_duplicates = df_raw[
    df_raw.duplicated(subset=["title", "author"], keep=False)
].sort_values(["title", "author"])

title_author_duplicates[["bookId", "title", "author", "isbn", "rating", "numRatings"]].head(20)

,bookId,title,author,isbn,rating,numRatings
2876,6443834-1q84,1Q84,"Haruki Murakami, 村上 春樹",9784103534228,4.00,49121
4463,7717482-1q84,1Q84,"Haruki Murakami, 村上 春樹",9784103534259,3.90,28602
1679,12177850-a-song-of-ice-and-fire,A Song of Ice and Fire,George R.R. Martin,9781780484259,4.56,53233
2899,9814682-a-song-of-ice-and-fire,A Song of Ice and Fire,George R.R. Martin,9780345529053,4.63,44894
37383,24633605-always-and-forever,Always and Forever,Harper Bentley (Goodreads Author),9999999999999,3.93,482
37437,24633605-always-and-forever,Always and Forever,Harper Bentley (Goodreads Author),9999999999999,3.93,482
37412,3973924-bal-masqu,Bal masqué,"Elia Barceló, Dorotea ter Horst (Translator)",9789021453293,3.79,341
37466,3973924-bal-masqu,Bal masqué,"Elia Barceló, Dorotea ter Horst (Translator)",9789021453293,3.79,341
37426,26845691-beatlebone,Beatlebone,Kevin Barry,9781782116134,3.47,2528
37480,26845691-beatlebone,Beatlebone,Kevin Barry,9781782116134,3.47,2528


Exact duplicate rows can likely be removed during cleaning because they repeat the same record.

However, `title + author` duplicates require caution. These may represent different editions, formats, translations, or duplicate scraped records. They should not be automatically dropped without further investigation.

#### 6. Type conversion audit

#### 7. List-column audit

#### 8. Date-column audit

#### 9. Author-column audit

#### 10. ISBN/pages/price audit

#### 11. Data quality decisions